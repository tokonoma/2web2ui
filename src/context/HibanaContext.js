import React, { createContext, useContext } from 'react';
import { connect } from 'react-redux';
import { updateUserUIOptions } from 'src/actions/currentUser';
import { showAlert } from 'src/actions/globalAlert';
import { isUserUiOptionSet } from 'src/helpers/conditions/user';
import { isAccountUiOptionSet } from 'src/helpers/conditions/account';
import { selectCondition } from 'src/selectors/accessConditionState';

const HibanaStateContext = createContext();

function Provider(props) {
  const { isHibanaEnabled: _isHibanaEnabled, children, ...rest } = props;

  // TODO: When OG theme is removed, the env variables will no longer be necessary
  const isHibanaEnabled =
    process.env.REACT_APP_DEFAULT_TO_HIBANA === 'true' ? true : _isHibanaEnabled;

  return (
    <HibanaStateContext.Provider value={{ isHibanaEnabled, ...rest }}>
      {children}
    </HibanaStateContext.Provider>
  );
}

function mapStateToProps(state) {
  return {
    isHibanaEnabled: selectCondition(isUserUiOptionSet('isHibanaEnabled'))(state),
    isBannerVisible: selectCondition(isUserUiOptionSet('isHibanaBannerVisible'))(state),
    hasThemeControls: selectCondition(isAccountUiOptionSet('hibana.hasThemeControls'))(state),
  };
}

const mapDispatchToProps = {
  dismissBanner: () => updateUserUIOptions({ isHibanaBannerVisible: false }),
  setIsHibanaEnabled: bool => {
    if (window.pendo && window.pendo.track) {
      window.pendo.track(`Hibana Toggle - ${Boolean(bool) ? 'On' : 'Off'}`);
    }

    // Always dismiss the banner when the user toggles their theme
    return updateUserUIOptions({ isHibanaEnabled: bool, isHibanaBannerVisible: false });
  },
  showAlert,
};

export function useHibana() {
  const context = useContext(HibanaStateContext);

  if (context === undefined) throw new Error('useHibana must be used within a HibanaProvider');

  return [context];
}

export const HibanaProvider = connect(mapStateToProps, mapDispatchToProps)(Provider);

export function HibanaConsumer({ children }) {
  return <HibanaStateContext.Consumer>{children}</HibanaStateContext.Consumer>;
}
