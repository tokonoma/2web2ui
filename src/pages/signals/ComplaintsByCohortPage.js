import React from 'react';
import { getComplaintsByCohort, getEngagementRecency } from 'src/actions/signals';
import { selectComplaintsByCohortDetails } from 'src/selectors/signals';
import { PageLink } from 'src/components/links';
import { Box, Grid, Panel, Stack } from 'src/components/matchbox';
import { useHibana } from 'src/context/HibanaContext';
import LineChart from './components/charts/linechart/LineChart';
import Legend from './components/charts/legend/Legend';
import Callout from 'src/components/callout';
import DateFilter from './components/filters/DateFilter';
import ComplaintsByCohortActions from './components/actionContent/ComplaintsByCohortActions';
import Page from './components/SignalsPage';
import Tabs from './components/engagement/Tabs';
import TooltipMetric from 'src/components/charts/TooltipMetric';
import withDetails from './containers/withDetails';
import withDateSelection from './containers/withDateSelection';
import { ENGAGEMENT_RECENCY_COHORTS } from './constants/info';
import { Loading } from 'src/components';
import { roundToPlaces } from 'src/helpers/units';
import moment from 'moment';
import _ from 'lodash';
import cohorts from './constants/cohorts';

export function ComplaintsByCohortPage(props) {
  const [state] = useHibana();
  const { isHibanaEnabled } = state;

  const isEmpty = () => {
    const { data } = props;

    // Returns true with 0 total complaints
    return data.every(({ p_total_fbl }) => !p_total_fbl);
  };

  const getYAxisProps = () => ({
    domain: isEmpty() ? [0, 1] : ['auto', 'auto'],
    tickFormatter: tick => `${roundToPlaces(tick * 100, 3)}%`,
  });

  const getXAxisProps = () => {
    const { xTicks } = props;

    return {
      ticks: xTicks,
      tickFormatter: tick => moment(tick).format('M/D'),
    };
  };

  const getTooltipContent = ({ payload = {} }) => {
    const metrics = _.keys(cohorts).reduce(
      (acc, key) => [
        ...acc,
        {
          ...cohorts[key],
          fill: isHibanaEnabled ? cohorts[key].hibanaFill : cohorts[key].OGFill,
          stroke: isHibanaEnabled ? cohorts[key].hibanaStroke : cohorts[key].OGStroke,
          key,
          value: payload[`p_${key}_fbl`],
        },
      ],
      [],
    );

    return (
      <Stack>
        {_.orderBy(metrics, 'value', 'desc').map(metric => (
          <TooltipMetric
            key={metric.key}
            color={metric.fill}
            label={metric.label}
            description={metric.description}
            value={`${roundToPlaces(metric.value * 100, 3)}%`}
          />
        ))}
      </Stack>
    );
  };

  const renderContent = () => {
    const {
      data = [],
      dataEngRecency = [],
      facet,
      facetId,
      handleDateSelect,
      loading,
      empty,
      error,
      selectedDate,
      shouldHighlightSelected,
      subaccountId,
    } = props;
    const selectedComplaints = _.find(data, ['date', selectedDate]) || {};
    const selectedEngagementRecency = _.find(dataEngRecency, ['date', selectedDate]) || {};
    let chartPanel;

    if (empty) {
      chartPanel = (
        <Callout title="No Data Available">Insufficient data to populate this chart</Callout>
      );
    }

    if (error) {
      chartPanel = <Callout title="Unable to Load Data">{error.message}</Callout>;
    }

    if (loading) {
      chartPanel = (
        <div style={{ height: '220px', position: 'relative' }}>
          <Loading />
        </div>
      );
    }

    return (
      <Grid>
        <Grid.Column sm={12} md={7}>
          <Panel data-id="complaint-rate-chart">
            <Tabs facet={facet} facetId={facetId} subaccountId={subaccountId} />
            <Panel.Section>
              {chartPanel || (
                <div className="LiftTooltip">
                  <LineChart
                    height={300}
                    onClick={handleDateSelect}
                    selected={selectedDate}
                    shouldHighlightSelected={shouldHighlightSelected}
                    lines={data}
                    tooltipWidth="250px"
                    tooltipContent={getTooltipContent}
                    yKeys={_.keys(cohorts).map(key => {
                      return {
                        key: `p_${key}_fbl`,
                        ...cohorts[key],
                        fill: isHibanaEnabled ? cohorts[key].hibanaFill : cohorts[key].OGFill,
                        stroke: isHibanaEnabled ? cohorts[key].hibanaStroke : cohorts[key].OGStroke,
                      };
                    })}
                    yAxisProps={getYAxisProps()}
                    xAxisProps={getXAxisProps()}
                  />
                  <Legend
                    items={_.values(cohorts)}
                    tooltipContent={label => ENGAGEMENT_RECENCY_COHORTS[label]}
                  />
                </div>
              )}
            </Panel.Section>
          </Panel>
        </Grid.Column>
        <Grid.Column sm={12} md={5} mdOffset={0}>
          <div>
            {!chartPanel && (
              <Box as={Panel} sectioned>
                <ComplaintsByCohortActions
                  complaintsByCohort={selectedComplaints}
                  recencyByCohort={selectedEngagementRecency}
                  date={selectedDate}
                />
              </Box>
            )}
          </div>
        </Grid.Column>
      </Grid>
    );
  };

  const { facet, facetId, subaccountId } = props;

  return (
    <Page
      breadcrumbAction={{
        content: 'Back to Engagement Recency Overview',
        to: '/signals/engagement',
        component: PageLink,
      }}
      title="Complaints by Cohort"
      facet={facet}
      facetId={facetId}
      subaccountId={subaccountId}
    >
      <Panel>
        <Panel.Section>
          <Grid>
            <Grid.Column xs={12} md={4}>
              <DateFilter label="Date Range" />
            </Grid.Column>
          </Grid>
        </Panel.Section>
      </Panel>
      {renderContent()}
    </Page>
  );
}

export default withDetails(
  withDateSelection(ComplaintsByCohortPage),
  { getComplaintsByCohort, getEngagementRecency },
  selectComplaintsByCohortDetails,
);
