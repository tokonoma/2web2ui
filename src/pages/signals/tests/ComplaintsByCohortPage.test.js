import React from 'react';
import { shallow } from 'enzyme';
import { useHibana } from 'src/context/HibanaContext';
import { ComplaintsByCohortPage } from '../ComplaintsByCohortPage';

jest.mock('src/context/HibanaContext');

describe('ComplaintsByCohortPage', () => {
  const data = [
    {
      date: '2017-01-01',
      p_total_fbl: 1,
    },
    {
      date: '2017-01-02',
      p_total_fbl: 10,
    },
  ];
  const dataEngRecency = [
    {
      date: '2017-01-01',
      c_uneng: 0.25,
    },
    {
      date: '2017-01-02',
      c_uneng: 0.5,
    },
  ];

  const defaultProps = {
    facetId: 'test.com',
    facet: 'sending-domain',
    handleDateSelect: jest.fn(),
    gap: 0.25,
    loading: false,
    empty: false,
    xTicks: [1, 2],
    selectedDate: '2017-01-02',
    shouldHighlightSelected: false,
    data,
    dataEngRecency,
  };

  const subject = props => {
    useHibana.mockImplementationOnce(() => [{ isHibanaEnabled: false }]);

    return shallow(<ComplaintsByCohortPage {...defaultProps} {...props} />);
  };

  it('renders correctly', () => {
    expect(subject()).toMatchSnapshot();
  });

  it('renders loading correctly', () => {
    const wrapper = subject({ loading: true });

    expect(wrapper.find('Panel')).toMatchSnapshot();
  });

  it('does not render SpamTrapsPreview when facet is mb_provider', () => {
    const wrapper = subject({ facet: 'mb_provider' });
    expect(
      wrapper.find('withRouter(Connect(WithDetails(SpamTrapsPreview)))'),
    ).not.toContainMatchingElement();
  });

  it('renders empty correctly', () => {
    const wrapper = subject({ empty: true });
    expect(wrapper.find('Panel')).toMatchSnapshot();
  });

  it('renders error correctly', () => {
    const wrapper = subject({ error: { message: 'error message' } });
    expect(wrapper.find('Callout').prop('children')).toEqual('error message');
    expect(wrapper.find('Callout').prop('title')).toEqual('Unable to Load Data');
  });

  describe('bar chart props', () => {
    it('renders tooltip content', () => {
      const wrapper = subject();
      const Tooltip = wrapper.find('LineChart').prop('tooltipContent');

      expect(
        shallow(
          <Tooltip
            payload={{
              p_uneng_fbl: 0.11111,
              p_365d_fbl: 0.2,
              p_90d_fbl: 0.3,
              p_14d_fbl: 0.4,
              p_new_fbl: 0.5,
              date: '2018-01-01',
              p_total_fbl: 10,
            }}
          />,
        ),
      ).toMatchSnapshot();
    });

    it('gets x axis props', () => {
      const wrapper = subject();
      const axisProps = wrapper.find('LineChart').prop('xAxisProps');

      expect(axisProps.ticks).toEqual([1, 2]);
      expect(axisProps.tickFormatter('2018-12-05')).toEqual('12/5');
    });

    it('gets y axis props with default domain', () => {
      const wrapper = subject({ data: [{ p_total_fbl: 0 }, { p_total_fbl: null }] });
      const axisProps = wrapper.find('LineChart').prop('yAxisProps');

      expect(axisProps.tickFormatter(0.252344)).toEqual('25.234%');
      expect(axisProps.domain).toEqual([0, 1]);
    });

    it('gets y axis props with domain', () => {
      const wrapper = subject({ data: [{ p_total_fbl: 0.5 }, { p_total_fbl: 0.6 }] });
      const axisProps = wrapper.find('LineChart').prop('yAxisProps');

      expect(axisProps.domain).toEqual(['auto', 'auto']);
    });
  });
});
