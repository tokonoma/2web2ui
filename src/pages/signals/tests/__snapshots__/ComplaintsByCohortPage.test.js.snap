// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComplaintsByCohortPage bar chart props renders tooltip content 1`] = `
<Stack>
  <TooltipMetric
    color="#88E2E7"
    description="Never engaged, first email in last 7 days"
    key="new"
    label="New"
    value="50%"
  />
  <TooltipMetric
    color="#1A838B"
    description="0-14 days since last engagement"
    key="14d"
    label="Recently Engaged"
    value="40%"
  />
  <TooltipMetric
    color="#219EA8"
    description="14-90 days since last engagement"
    key="90d"
    label="Semi Recently Engaged"
    value="30%"
  />
  <TooltipMetric
    color="#29B9C7"
    description="90-365 days since last engagement"
    key="365d"
    label="Not Recently Engaged"
    value="20%"
  />
  <TooltipMetric
    color="#50CFDA"
    description="365+ days since last engagement"
    key="uneng"
    label="Never Engaged"
    value="11.111%"
  />
</Stack>
`;

exports[`ComplaintsByCohortPage renders correctly 1`] = `
<SignalsPage
  breadcrumbAction={
    Object {
      "component": [Function],
      "content": "Back to Engagement Recency Overview",
      "to": "/signals/engagement",
    }
  }
  facet="sending-domain"
  facetId="test.com"
  title="Complaints by Cohort"
>
  <Panel>
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          xs={12}
        >
          <Connect(DateFilter)
            label="Date Range"
          />
        </Grid.Column>
      </Grid>
    </Panel.Section>
  </Panel>
  <Grid>
    <Grid.Column
      md={7}
      sm={12}
    >
      <Panel
        data-id="complaint-rate-chart"
      >
        <withRouter(Tabs)
          facet="sending-domain"
          facetId="test.com"
        />
        <Panel.Section>
          <div
            className="LiftTooltip"
          >
            <LineChart
              height={300}
              lineType="linear"
              lines={
                Array [
                  Object {
                    "date": "2017-01-01",
                    "p_total_fbl": 1,
                  },
                  Object {
                    "date": "2017-01-02",
                    "p_total_fbl": 10,
                  },
                ]
              }
              margin={
                Object {
                  "bottom": 5,
                  "left": 18,
                  "right": 0,
                  "top": 12,
                }
              }
              onClick={[MockFunction]}
              selected="2017-01-02"
              shouldHighlightSelected={false}
              stroke="#000"
              tooltipContent={[Function]}
              tooltipWidth="250px"
              width="99%"
              xAxisProps={
                Object {
                  "tickFormatter": [Function],
                  "ticks": Array [
                    1,
                    2,
                  ],
                }
              }
              xKey="date"
              yAxisProps={
                Object {
                  "domain": Array [
                    "auto",
                    "auto",
                  ],
                  "tickFormatter": [Function],
                }
              }
              yKey="value"
              yKeys={
                Array [
                  Object {
                    "OGFill": "#88E2E7",
                    "OGStroke": "#88E2E7",
                    "description": "Never engaged, first email in last 7 days",
                    "fill": "#88E2E7",
                    "hibanaFill": "#78b6ff",
                    "hibanaStroke": "#78b6ff",
                    "key": "p_new_fbl",
                    "label": "New",
                    "stroke": "#88E2E7",
                  },
                  Object {
                    "OGFill": "#50CFDA",
                    "OGStroke": "#50CFDA",
                    "description": "365+ days since last engagement",
                    "fill": "#50CFDA",
                    "hibanaFill": "#429bff",
                    "hibanaStroke": "#429bff",
                    "key": "p_uneng_fbl",
                    "label": "Never Engaged",
                    "stroke": "#50CFDA",
                  },
                  Object {
                    "OGFill": "#29B9C7",
                    "OGStroke": "#29B9C7",
                    "description": "90-365 days since last engagement",
                    "fill": "#29B9C7",
                    "hibanaFill": "#1273e6",
                    "hibanaStroke": "#1273e6",
                    "key": "p_365d_fbl",
                    "label": "Not Recently Engaged",
                    "stroke": "#29B9C7",
                  },
                  Object {
                    "OGFill": "#219EA8",
                    "OGStroke": "#219EA8",
                    "description": "14-90 days since last engagement",
                    "fill": "#219EA8",
                    "hibanaFill": "#0060cc",
                    "hibanaStroke": "#0060cc",
                    "key": "p_90d_fbl",
                    "label": "Semi Recently Engaged",
                    "stroke": "#219EA8",
                  },
                  Object {
                    "OGFill": "#1A838B",
                    "OGStroke": "#1A838B",
                    "description": "0-14 days since last engagement",
                    "fill": "#1A838B",
                    "hibanaFill": "#1a497d",
                    "hibanaStroke": "#1a497d",
                    "key": "p_14d_fbl",
                    "label": "Recently Engaged",
                    "stroke": "#1A838B",
                  },
                ]
              }
              yRange={
                Array [
                  "auto",
                  "auto",
                ]
              }
            />
            <Legend
              items={
                Array [
                  Object {
                    "OGFill": "#88E2E7",
                    "OGStroke": "#88E2E7",
                    "description": "Never engaged, first email in last 7 days",
                    "hibanaFill": "#78b6ff",
                    "hibanaStroke": "#78b6ff",
                    "label": "New",
                  },
                  Object {
                    "OGFill": "#50CFDA",
                    "OGStroke": "#50CFDA",
                    "description": "365+ days since last engagement",
                    "hibanaFill": "#429bff",
                    "hibanaStroke": "#429bff",
                    "label": "Never Engaged",
                  },
                  Object {
                    "OGFill": "#29B9C7",
                    "OGStroke": "#29B9C7",
                    "description": "90-365 days since last engagement",
                    "hibanaFill": "#1273e6",
                    "hibanaStroke": "#1273e6",
                    "label": "Not Recently Engaged",
                  },
                  Object {
                    "OGFill": "#219EA8",
                    "OGStroke": "#219EA8",
                    "description": "14-90 days since last engagement",
                    "hibanaFill": "#0060cc",
                    "hibanaStroke": "#0060cc",
                    "label": "Semi Recently Engaged",
                  },
                  Object {
                    "OGFill": "#1A838B",
                    "OGStroke": "#1A838B",
                    "description": "0-14 days since last engagement",
                    "hibanaFill": "#1a497d",
                    "hibanaStroke": "#1a497d",
                    "label": "Recently Engaged",
                  },
                ]
              }
              tooltipContent={[Function]}
            />
          </div>
        </Panel.Section>
      </Panel>
    </Grid.Column>
    <Grid.Column
      md={5}
      mdOffset={0}
      sm={12}
    >
      <div>
        <Box
          as={[Function]}
          sectioned={true}
        >
          <ComplaintsByCohortActions
            complaintsByCohort={
              Object {
                "date": "2017-01-02",
                "p_total_fbl": 10,
              }
            }
            date="2017-01-02"
            recencyByCohort={
              Object {
                "c_uneng": 0.5,
                "date": "2017-01-02",
              }
            }
          />
        </Box>
      </div>
    </Grid.Column>
  </Grid>
</SignalsPage>
`;

exports[`ComplaintsByCohortPage renders empty correctly 1`] = `
Array [
  <Panel>
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          xs={12}
        >
          <Connect(DateFilter)
            label="Date Range"
          />
        </Grid.Column>
      </Grid>
    </Panel.Section>
  </Panel>,
  <Panel
    data-id="complaint-rate-chart"
  >
    <withRouter(Tabs)
      facet="sending-domain"
      facetId="test.com"
    />
    <Panel.Section>
      <Callout
        title="No Data Available"
      >
        Insufficient data to populate this chart
      </Callout>
    </Panel.Section>
  </Panel>,
]
`;

exports[`ComplaintsByCohortPage renders loading correctly 1`] = `
Array [
  <Panel>
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          xs={12}
        >
          <Connect(DateFilter)
            label="Date Range"
          />
        </Grid.Column>
      </Grid>
    </Panel.Section>
  </Panel>,
  <Panel
    data-id="complaint-rate-chart"
  >
    <withRouter(Tabs)
      facet="sending-domain"
      facetId="test.com"
    />
    <Panel.Section>
      <div
        style={
          Object {
            "height": "220px",
            "position": "relative",
          }
        }
      >
        <Loading />
      </div>
    </Panel.Section>
  </Panel>,
]
`;
